### PR TITLE
Make `FlagAssignment` an opaque `newtype`

### PR DESCRIPTION
**Pull request by hvr**
This is a refactoring abstracting `FlagAssignment` while retaining its
external appearance as much as possible (i.e. same Read/Show/Binary
instances etc).

Later we can attach new instances, enforce internal invariants (like
e.g. uniqueness of flagnames), switch out the internal
representation (maybe to `Data.Map`), etc more easily.

Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog.
* [x] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

File: cabal-install/Distribution/Client/BuildReports/Anonymous.hs [hunk 1]
```diff
@@ -26,6 +26,9 @@ module Distribution.Client.BuildReports.Anonymous (
 --    showList,
   ) where
 
+import Prelude (mapM)
```
Comment (phadej) [inline cabal-install/Distribution/Client/BuildReports/Anonymous.hs]: why not `traverse`?

File: cabal-install/Distribution/Client/GenBounds.hs [hunk 2]
```diff
@@ -15,6 +15,9 @@ module Distribution.Client.GenBounds (
     genBounds
   ) where
 
+import Prelude (mapM_)
```
Comment (phadej) [inline cabal-install/Distribution/Client/GenBounds.hs]: and `traverse_`

File: cabal-install/Distribution/Client/BuildReports/Anonymous.hs [hunk 3]
```diff
@@ -267,7 +265,7 @@ sortedFieldDescrs = sortBy (comparing fieldName) fieldDescrs
 
 dispFlag :: (FlagName, Bool) -> Disp.Doc
 dispFlag (fname, True)  =                  Disp.text (unFlagName fname)
-dispFlag (fname, False) = Disp.char '-' <> Disp.text (unFlagName fname)
+dispFlag (fname, False) = Disp.char '-' PP.<> Disp.text (unFlagName fname)
```
Comment (phadej) [inline cabal-install/Distribution/Client/BuildReports/Anonymous.hs]: we have `<<>>`, but I don't insist.

File: Cabal/Distribution/Simple/Configure.hs [hunk 4]
```diff
@@ -922,10 +922,10 @@ configureFinalizedPackage verbosity cfg enabled
     -- we do it here so that those get checked too
     let pkg_descr = addExtraIncludeLibDirs pkg_descr0'
 
-    when (not (null flags)) $
+    when (flags /= mempty) $
```
Comment (phadej) [inline Cabal/Distribution/Simple/Configure.hs]: should we have `nullFlagAssignment`?

File: cabal-install/Distribution/Client/Dependency.hs [hunk 5]
```diff
@@ -898,7 +898,7 @@ configuredPackageProblems :: Platform -> CompilerInfo
                           -> SolverPackage UnresolvedPkgLoc -> [PackageProblem]
 configuredPackageProblems platform cinfo
   (SolverPackage pkg specifiedFlags stanzas specifiedDeps' _specifiedExeDeps') =
-     [ DuplicateFlag flag | ((flag,_):_) <- duplicates specifiedFlags ]
+     [ DuplicateFlag flag | ((flag,_):_) <- duplicates (PD.unFlagAssignment specifiedFlags) ] -- FIXME/TODO
```
Comment (phadej) [inline cabal-install/Distribution/Client/Dependency.hs]: could you comment a little more context to TODO

File: cabal-install/Distribution/Client/Install.hs [hunk 6]
```diff
@@ -419,7 +418,7 @@ planPackages verbosity comp platform mSandboxPkgInfo solver
                      (PackagePropertyFlags flags)
             in LabeledPackageConstraint pc ConstraintSourceConfigFlagOrTarget
           | let flags = configConfigurationsFlags configFlags
-          , not (null flags)
+          , flags /= mempty
```
Comment (phadej) [inline cabal-install/Distribution/Client/Install.hs]: `nullFlagAssignment`

**Comment by hvr** [comment]
fwiw, turning `FlagAssignment` into a `Map FlagName Bool` will require a bit of thought, as we need to figure out how to handle the situation when conflicting flag settings collide, as right now the checks for inconsistent flag assignments are deferred.

**Pull request merged**